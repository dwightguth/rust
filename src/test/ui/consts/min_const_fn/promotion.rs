#![feature(min_const_fn)]

use std::cell::Cell;

const fn foo1() {}
const fn foo2(x: i32) -> i32 { x }
const fn foo3() -> i32 { 42 }
const fn foo4() -> Cell<i32> { Cell::new(42) }
const fn foo5() -> Option<Cell<i32>> { Some(Cell::new(42)) }
const fn foo6() -> Option<Cell<i32>> { None }

fn main() {
    let x: &'static () = &foo1();
    let y: &'static i32 = &foo2(42); //~ ERROR does not live long enough
    let z: &'static i32 = &foo3();
    let a: &'static Cell<i32> = &foo4(); // doesn't error on HIR borrowck
    let a: &'static Option<Cell<i32>> = &foo5(); // doesn't error on HIR borrowck
    let a: &'static Option<Cell<i32>> = &foo6(); // doesn't error on HIR borrowck
}
