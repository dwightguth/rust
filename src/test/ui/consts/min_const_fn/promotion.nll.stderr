error[E0597]: borrowed value does not live long enough
  --> $DIR/promotion.rs:14:28
   |
LL |     let y: &'static i32 = &foo2(42); //~ ERROR does not live long enough
   |                            ^^^^^^^^ temporary value does not live long enough
...
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promotion.rs:16:34
   |
LL |     let a: &'static Cell<i32> = &foo4();
   |                                  ^^^^^^ temporary value does not live long enough
...
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promotion.rs:17:42
   |
LL |     let a: &'static Option<Cell<i32>> = &foo5();
   |                                          ^^^^^^ temporary value does not live long enough
LL |     let a: &'static Option<Cell<i32>> = &foo6();
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promotion.rs:18:42
   |
LL |     let a: &'static Option<Cell<i32>> = &foo6();
   |                                          ^^^^^^ temporary value does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0597`.
